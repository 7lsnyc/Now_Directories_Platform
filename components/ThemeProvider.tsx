import { ReactNode } from 'react';
import { DEFAULT_THEME_COLORS } from '@/constants/theme';

interface ThemeColors {
  primary: string;
  secondary: string;
  accent: string;
  primaryText?: string;
  secondaryText?: string;
  accentText?: string;
}

export interface ThemeProviderProps {
  children: ReactNode;
  directory: string;
  themeColors?: ThemeColors;
  config?: any; // Support for backward compatibility with tests
}

/**
 * ThemeProvider component
 * Sets CSS variables for the directory theme colors
 * For multi-domain architecture, this applies brand colors from the Supabase directories table
 */
export default function ThemeProvider({ children, directory, themeColors, config }: ThemeProviderProps) {
  // For backward compatibility with tests that pass config
  const directoryName = config?.name || directory;
  
  // Use themeColors from config if provided (for test backward compatibility)
  const configThemeColors = config?.theme?.colors;
  const effectiveThemeColors = themeColors || configThemeColors || DEFAULT_THEME_COLORS;
  
  // Use provided themeColors or fallback to defaults
  const { 
    primary = DEFAULT_THEME_COLORS.primary, 
    secondary = DEFAULT_THEME_COLORS.secondary, 
    accent = DEFAULT_THEME_COLORS.accent, 
    primaryText = DEFAULT_THEME_COLORS.primaryText, 
    secondaryText = DEFAULT_THEME_COLORS.secondaryText, 
    accentText = DEFAULT_THEME_COLORS.accentText 
  } = effectiveThemeColors;
  
  return (
    <div 
      className={`theme-provider theme-${directoryName}`}
      data-testid="theme-provider"
      data-directory={directoryName}
      style={{
        // Set theme colors as CSS variables
        '--color-primary': primary,
        '--color-secondary': secondary,
        '--color-accent': accent,
        '--color-primary-text': primaryText,
        '--color-secondary-text': secondaryText,
        '--color-accent-text': accentText,
        // Set theme gradient
        '--gradient-primary': `linear-gradient(135deg, ${primary} 0%, ${secondary} 100%)`,
      } as React.CSSProperties}
    >
      {children}
    </div>
  );
}
