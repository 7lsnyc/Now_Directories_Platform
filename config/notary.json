{
  "name": "notary",
  "title": "Notary Finder Now",
  "description": "Find qualified notaries in your area offering mobile, 24-hour, and free services.",
  "logo": {
    "path": "/logos/notary-logo.svg",
    "alt": "Notary Finder Now"
  },
  "theme": {
    "name": "blue-notary",
    "colors": {
      "primary": "#1e40af",
      "secondary": "#1e3a8a",
      "accent": "#f97316"
    }
  },
  "hero": {
    "heading": "Find a Qualified Notary Near You â€” Now!",
    "subheading": "Connect with mobile, 24-hour, and free notary services in your area instantly."
  },
  "serviceTypes": [
    "Mobile Notaries",
    "24-Hour Availability",
    "Remote Notaries",
    "Free Services"
  ],
  "navigation": {
    "header": {
      "ctaButton": {
        "text": "Request Featured Listing",
        "url": "/request-listing"
      }
    },
    "footer": {
      "quickLinks": [
        { "text": "Home", "url": "/" },
        { "text": "Find a Notary", "url": "/search" },
        { "text": "About Us", "url": "/about" },
        { "text": "Contact", "url": "/contact" }
      ],
      "services": [
        { "text": "Mobile Notaries", "url": "/services/mobile" },
        { "text": "24-Hour Availability", "url": "/services/24-hour" },
        { "text": "Remote Notaries", "url": "/services/remote" },
        { "text": "Free Services", "url": "/services/free" }
      ],
      "support": [
        { "text": "FAQ", "url": "/faq" },
        { "text": "Privacy Policy", "url": "/privacy" },
        { "text": "Terms of Service", "url": "/terms" },
        { "text": "Request Featured Listing", "url": "/request-listing" }
      ]
    }
  },
  "seo": {
    "title": "Notary Finder Now - Find Qualified Mobile Notaries Near You",
    "description": "Connect with mobile, 24-hour, and free notary services in your area. Trusted notaries available when you need them."
  },
  "features": {
    "search": true,
    "filter": true,
    "sort": true,
    "pagination": true
  }
}