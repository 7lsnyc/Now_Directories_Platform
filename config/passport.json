{
  "name": "passport",
  "title": "Passport Services Directory",
  "description": "Find passport expediting services, photo locations, and application assistance near you.",
  "logo": {
    "path": "/logos/passport-logo.svg",
    "alt": "Passport Services Directory"
  },
  "theme": {
    "name": "green-passport",
    "colors": {
      "primary": "#15803d",
      "secondary": "#166534",
      "accent": "#ea580c"
    }
  },
  "hero": {
    "heading": "Get Your Passport Fast â€” Find Services Near You",
    "subheading": "Connect with expedited passport services, photo locations, and application assistance in your area."
  },
  "serviceTypes": [
    "Expedited Processing",
    "Passport Photos",
    "Application Assistance",
    "Renewal Services"
  ],
  "navigation": {
    "header": {
      "ctaButton": {
        "text": "Add Your Service",
        "url": "/add-service"
      }
    },
    "footer": {
      "quickLinks": [
        { "text": "Home", "url": "/" },
        { "text": "Find Services", "url": "/search" },
        { "text": "About", "url": "/about" },
        { "text": "Contact", "url": "/contact" }
      ],
      "services": [
        { "text": "Expedited Processing", "url": "/services/expedited" },
        { "text": "Passport Photos", "url": "/services/photos" },
        { "text": "Application Assistance", "url": "/services/assistance" },
        { "text": "Renewal Services", "url": "/services/renewal" }
      ],
      "support": [
        { "text": "FAQ", "url": "/faq" },
        { "text": "Privacy Policy", "url": "/privacy" },
        { "text": "Terms of Service", "url": "/terms" },
        { "text": "Add Your Service", "url": "/add-service" }
      ]
    }
  },
  "seo": {
    "title": "Passport Services Directory - Find Fast Passport Services Near You",
    "description": "Find passport expediting services, photo locations, and application assistance in your area for fast, reliable passport processing."
  },
  "features": {
    "search": true,
    "filter": true,
    "sort": true,
    "pagination": true
  }
}
