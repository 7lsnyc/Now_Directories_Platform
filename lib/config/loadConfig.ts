import fs from 'fs';
import path from 'path';
import { createClient } from '@/lib/supabase/server';
import { cache } from 'react';

export interface DirectoryConfig {
  name: string;
  title: string;
  description: string;
  logo: {
    path: string;
    alt: string;
  };
  theme: {
    name: string;
    colors: {
      primary: string;
      secondary: string;
      accent: string;
    };
  };
  hero: {
    heading: string;
    subheading: string;
  };
  serviceTypes: string[];
  navigation: {
    header: {
      ctaButton: {
        text: string;
        url: string;
      };
    };
    footer: {
      quickLinks: Array<{ text: string; url: string }>;
      services: Array<{ text: string; url: string }>;
      support: Array<{ text: string; url: string }>;
    };
  };
  seo: {
    title: string;
    description: string;
  };
  features: {
    search: boolean;
    filter: boolean;
    sort: boolean;
    pagination: boolean;
  };
}

/**
 * Default configuration used as fallback when a directory config is not found
 */
export const defaultConfig: DirectoryConfig = {
  name: 'default',
  title: 'Directory Services',
  description: 'Find local services in your area.',
  logo: {
    path: '/logos/default-logo.svg',
    alt: 'Directory Services'
  },
  theme: {
    name: 'default-theme',
    colors: {
      primary: '#3b82f6',
      secondary: '#2563eb',
      accent: '#f59e0b'
    }
  },
  hero: {
    heading: 'Find Local Services',
    subheading: 'Discover trusted providers in your area.'
  },
  serviceTypes: [
    'Service Type 1',
    'Service Type 2',
    'Service Type 3',
    'Service Type 4'
  ],
  navigation: {
    header: {
      ctaButton: {
        text: 'Add Listing',
        url: '/add-listing'
      }
    },
    footer: {
      quickLinks: [
        { text: 'Home', url: '/' },
        { text: 'Search', url: '/search' },
        { text: 'About', url: '/about' },
        { text: 'Contact', url: '/contact' }
      ],
      services: [
        { text: 'Service 1', url: '/services/1' },
        { text: 'Service 2', url: '/services/2' },
        { text: 'Service 3', url: '/services/3' },
        { text: 'Service 4', url: '/services/4' }
      ],
      support: [
        { text: 'FAQ', url: '/faq' },
        { text: 'Privacy', url: '/privacy' },
        { text: 'Terms', url: '/terms' },
        { text: 'Add Listing', url: '/add-listing' }
      ]
    }
  },
  seo: {
    title: 'Directory Services - Find Local Providers',
    description: 'Find local service providers in your area. Trusted professionals available when you need them.'
  },
  features: {
    search: true,
    filter: true,
    sort: true,
    pagination: true
  }
};

/**
 * Loads configuration for a specific directory by slug
 * This function prioritizes fetching data from Supabase as the source of truth
 * @param slug - The directory slug to load configuration for
 * @returns The directory configuration or default config if not found
 */
export const loadConfig = cache(async function(slug?: string): Promise<DirectoryConfig> {
  // If no slug is provided, return the default config
  if (!slug) {
    return defaultConfig;
  }

  try {
    // First attempt to get the directory config from Supabase (source of truth)
    const supabase = createClient();
    const { data, error } = await supabase
      .from('directories')
      .select('*')
      .eq('directory_slug', slug)
      .eq('is_active', true)
      .single();
    
    if (error) {
      console.error(`Error fetching directory from Supabase for slug ${slug}:`, error);
    }
    
    // If we successfully retrieved the directory data from Supabase, convert it to a DirectoryConfig
    if (data) {
      console.log(`Found directory in Supabase for slug: ${slug}`);
      
      // Convert the Supabase directory data to DirectoryConfig format
      // This assumes certain fields are present in the Supabase data
      const config: DirectoryConfig = {
        name: data.name || defaultConfig.name,
        title: data.title || defaultConfig.title,
        description: data.description || defaultConfig.description,
        logo: {
          path: data.logo_path || defaultConfig.logo.path,
          alt: data.logo_alt || defaultConfig.logo.alt
        },
        theme: {
          name: data.theme_name || defaultConfig.theme.name,
          colors: {
            primary: data.brand_color_primary || defaultConfig.theme.colors.primary,
            secondary: data.brand_color_secondary || defaultConfig.theme.colors.secondary,
            accent: data.brand_color_accent || defaultConfig.theme.colors.accent
          }
        },
        hero: {
          heading: data.hero_heading || defaultConfig.hero.heading,
          subheading: data.hero_subheading || defaultConfig.hero.subheading
        },
        serviceTypes: data.service_types || defaultConfig.serviceTypes,
        navigation: data.navigation || defaultConfig.navigation,
        seo: {
          title: data.seo_title || defaultConfig.seo.title,
          description: data.seo_description || defaultConfig.seo.description
        },
        features: data.features || defaultConfig.features
      };
      
      return config;
    }
    
    // If Supabase lookup failed, log a warning
    console.warn(`Directory not found in Supabase for slug: ${slug}`);
    
    // For development only: Fall back to local config if absolutely necessary
    // In a production environment, this should return defaultConfig
    return defaultConfig;
  } catch (error) {
    console.error(`Error loading config for slug ${slug}:`, error);
    return defaultConfig;
  }
});