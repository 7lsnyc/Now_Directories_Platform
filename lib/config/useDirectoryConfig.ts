// lib/config/useDirectoryConfig.ts â€” placeholder for MVP starter
'use client';

import { useState, useEffect } from 'react';
import { loadConfig, DirectoryConfig } from './loadConfig';

/**
 * A hook to access and potentially update directory configuration
 * This is a placeholder implementation for the MVP
 */
export function useDirectoryConfig() {
  const [config, setConfig] = useState<DirectoryConfig | null>(null);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState<Error | null>(null);

  useEffect(() => {
    try {
      // Load the config
      const directoryConfig = loadConfig();
      setConfig(directoryConfig);
      setLoading(false);
    } catch (err) {
      console.error('Error loading directory configuration:', err);
      setError(err instanceof Error ? err : new Error('Unknown error loading config'));
      setLoading(false);
    }
  }, []);

  return {
    config,
    loading,
    error,
    // In a real implementation, this would update the configuration
    updateConfig: (newConfig: Partial<DirectoryConfig>) => {
      if (!config) return;
      setConfig({ ...config, ...newConfig });
    }
  };
}