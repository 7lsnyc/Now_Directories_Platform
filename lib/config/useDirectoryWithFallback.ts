// lib/config/useDirectoryWithFallback.ts â€” placeholder for MVP starter
'use client';

import { useState, useEffect } from 'react';
import { DirectoryConfig } from './loadConfig';
import { useDirectoryConfig } from './useDirectoryConfig';

/**
 * Hook that provides directory configuration with fallback values
 * Ensures critical values are always available even if the main config fails to load
 */
export function useDirectoryWithFallback() {
  const { config, loading, error } = useDirectoryConfig();
  const [fallbackConfig] = useState<DirectoryConfig>({
    title: 'Directory',
    description: 'Default directory view',
    theme: {
      primary: '#4B5563', // gray-600
      secondary: '#1F2937', // gray-800
      accent: '#F59E0B', // amber-500
    },
    features: {
      search: true,
      filter: false,
      sort: true,
      pagination: true,
    },
  });

  // Merge loaded config with fallback config, ensuring critical values
  const safeConfig = config ? {
    ...fallbackConfig,
    ...config,
    // Ensure nested objects are also merged properly
    theme: {
      ...fallbackConfig.theme,
      ...(config?.theme || {}),
    },
    features: {
      ...fallbackConfig.features,
      ...(config?.features || {}),
    },
  } : fallbackConfig;

  return {
    config: safeConfig,
    loading,
    error,
    isFallback: !config || !!error,
  };
}