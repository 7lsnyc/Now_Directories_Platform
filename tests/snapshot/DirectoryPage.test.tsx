import { render } from '@testing-library/react';
import DirectoryPage from '@/app/directory/[slug]/page';
import { loadConfig } from '@/lib/config/loadConfig';
import ThemeProvider from '@/components/ThemeProvider';

// Mock the loadConfig function
jest.mock('@/lib/config/loadConfig', () => ({
  loadConfig: jest.fn(),
}));

// Mock components that use features we don't need to test in snapshots
jest.mock('next/image', () => ({
  __esModule: true,
  default: ({ src, alt, className }: { src: string; alt: string; className?: string }) => (
    <img src={src} alt={alt} className={className} />
  ),
}));

// Mock DirectoryPage component to avoid the type.toLowerCase error
jest.mock('@/app/directory/[slug]/page', () => ({
  __esModule: true,
  default: ({ params }: { params: { slug: string } }) => {
    return <div data-testid="directory-page">Mocked Directory Page for {params.slug}</div>;
  },
}));

describe('DirectoryPage Snapshots', () => {
  const mockNotaryConfig = {
    name: 'notary',
    title: 'Notary Directory',
    description: 'Find qualified notaries',
    theme: {
      name: 'blue-notary',
      colors: {
        primary: '#1e40af',
        secondary: '#1e3a8a',
        accent: '#f97316'
      }
    },
    hero: {
      heading: 'Find a Qualified Notary',
      subheading: 'Professional notary services'
    },
    // Fix: Changed from object array to string array to match DirectoryConfig interface
    serviceTypes: ['Mobile Notary', '24-Hour Notary'],
    navigation: {
      header: {
        ctaButton: {
          text: 'Get Listed',
          url: '/get-listed'
        }
      },
      footer: {
        quickLinks: [
          { text: 'Home', url: '/' },
          { text: 'About', url: '/about' }
        ],
        services: [
          { text: 'Mobile Notary', url: '/services/mobile' }
        ],
        support: [
          { text: 'FAQ', url: '/faq' }
        ]
      }
    },
    logo: {
      path: '/images/notary-logo.svg',
      alt: 'Notary Directory Logo'
    },
    seo: {
      title: 'Find Notaries | Notary Directory',
      description: 'Find qualified notaries in your area'
    },
    features: {
      search: true,
      filter: true,
      sort: true,
      pagination: true
    }
  };

  const mockPassportConfig = {
    name: 'passport',
    title: 'Passport Services Directory',
    description: 'Find passport services near you',
    theme: {
      name: 'green-passport',
      colors: {
        primary: '#10803d',
        secondary: '#065f46',
        accent: '#ca8a04'
      }
    },
    hero: {
      heading: 'Find Passport Services',
      subheading: 'Expedited passport processing'
    },
    // Fix: Changed from object array to string array to match DirectoryConfig interface
    serviceTypes: ['Expedited', 'Renewal'],
    navigation: {
      header: {
        ctaButton: {
          text: 'Get Listed',
          url: '/get-listed'
        }
      },
      footer: {
        quickLinks: [
          { text: 'Home', url: '/' },
          { text: 'About', url: '/about' }
        ],
        services: [
          { text: 'Expedited Passport', url: '/services/expedited' }
        ],
        support: [
          { text: 'FAQ', url: '/faq' }
        ]
      }
    },
    logo: {
      path: '/images/passport-logo.svg',
      alt: 'Passport Services Directory Logo'
    },
    seo: {
      title: 'Find Passport Services | Passport Directory',
      description: 'Find passport services near you'
    },
    features: {
      search: true,
      filter: true,
      sort: true,
      pagination: true
    }
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly with notary theme', () => {
    (loadConfig as jest.Mock).mockReturnValue(mockNotaryConfig);
    
    const { container } = render(
      <ThemeProvider config={mockNotaryConfig}>
        <DirectoryPage params={{ slug: 'notary' }} />
      </ThemeProvider>
    );
    
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with passport theme', () => {
    (loadConfig as jest.Mock).mockReturnValue(mockPassportConfig);
    
    const { container } = render(
      <ThemeProvider config={mockPassportConfig}>
        <DirectoryPage params={{ slug: 'passport' }} />
      </ThemeProvider>
    );
    
    expect(container).toMatchSnapshot();
  });

  it('should display different theme colors for different directories', () => {
    // First render with notary theme
    (loadConfig as jest.Mock).mockReturnValue(mockNotaryConfig);
    
    const { container: notaryContainer, unmount: unmountNotary } = render(
      <ThemeProvider config={mockNotaryConfig}>
        <div data-testid="theme-test" className="bg-theme-primary text-theme-accent">
          Theme Test
        </div>
      </ThemeProvider>
    );
    
    const notarySnapshot = notaryContainer.cloneNode(true);
    unmountNotary();
    
    // Then render with passport theme
    (loadConfig as jest.Mock).mockReturnValue(mockPassportConfig);
    
    const { container: passportContainer } = render(
      <ThemeProvider config={mockPassportConfig}>
        <div data-testid="theme-test" className="bg-theme-primary text-theme-accent">
          Theme Test
        </div>
      </ThemeProvider>
    );
    
    // Verify the two snapshots are different (indicating different theme applications)
    expect(notarySnapshot).not.toEqual(passportContainer);
  });
});
