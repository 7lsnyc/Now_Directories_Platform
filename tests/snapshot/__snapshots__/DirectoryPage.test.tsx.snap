// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DirectoryPage Snapshots should render correctly with notary theme 1`] = `
<div>
  <div
    class="theme-provider theme-notary"
    data-directory="notary"
    data-testid="theme-provider"
    style="--color-primary: #1e40af; --color-secondary: #1e3a8a; --color-accent: #f97316; --color-primary-text: #FFFFFF; --color-secondary-text: #FFFFFF; --color-accent-text: #000000; --gradient-primary: linear-gradient(135deg, #1e40af 0%, #1e3a8a 100%);"
  >
    <div
      data-testid="directory-page"
    >
      Mocked Directory Page for 
      notary
    </div>
  </div>
</div>
`;

exports[`DirectoryPage Snapshots should render correctly with passport theme 1`] = `
<div>
  <div
    class="theme-provider theme-passport"
    data-directory="passport"
    data-testid="theme-provider"
    style="--color-primary: #10803d; --color-secondary: #065f46; --color-accent: #ca8a04; --color-primary-text: #FFFFFF; --color-secondary-text: #FFFFFF; --color-accent-text: #000000; --gradient-primary: linear-gradient(135deg, #10803d 0%, #065f46 100%);"
  >
    <div
      data-testid="directory-page"
    >
      Mocked Directory Page for 
      passport
    </div>
  </div>
</div>
`;
